abmarl/policies/q_table_policy.py

```diff
@@ -0,0 +1,156 @@
+
+    @q_table.setter
+    def q_table(self, value):
+        assert isinstance(value, np.ndarray), "Q-Table must be a numpy array."
+        assert len(value.shape) == 2, "Q-Table must have two dimensions."
+        assert value.shape[0] == self.observation_space.n, \
+            "The number of rows in the q table must be the same as the size of " + \
+            "the observation space."
+        assert value.shape[1] == self.action_space.n, \
+            "The number of columns in the q table must be the same as the size of " + \
+            "the action space."
+        self._q_table = value
+
+    @property
+    def action_space(self):
+        """
+        The space of possible output actions for this policy.
+
+        Must be Discrete.
+        """
+        return self._action_space
+
+    @action_space.setter
+    def action_space(self, value):
+        assert isinstance(value, Discrete), "Action space must be Discrete."
+        self._action_space = value
+
+    @property
+    def observation_space(self):
+        """
+        The space of possible input observations for this policy.
+
+        Must be Discrete.
+        """
+        return self._observation_space
+
+    @observation_space.setter
+    def observation_space(self, value):
+        assert isinstance(value, Discrete), "Observation space must be Discrete."
+        self._observation_space = value
+
+    @classmethod
+    def build(cls, policy_or_table, **kwargs):
+        """
+        Build a policy from another policy or q_table.
+
+        Args:
+            policy_or_table: We can create a policy using another policy, in which
+                case we look at the policy's q_table. Or we can create a policy
+                from the q_table directly.
+
+        Returns:
+            A new policy with the q_table.
+        """
+        if isinstance(policy_or_table, QTablePolicy):
+            q_table = policy_or_table.q_table
+        else:
+            q_table = policy_or_table
+        new_policy = cls(
+            observation_space=Discrete(q_table.shape[0]),
+            action_space=Discrete(q_table.shape[1]),
+            **kwargs
+        )
+        new_policy.q_table = q_table
+        return new_policy
+
+    @abstractmethod
+    def probability(self, obs, action, **kwargs):
+        """
+        Calculate the probability of choosing this action given this observation.
+        """
+        pass
+
+    def update(self, obs, action, value, **kwargs):
+        """
+        Set the q_table at (obs, action) to the value.
+        """
+        self.q_table[obs, action] = value
+
+
+class GreedyPolicy(QTablePolicy):
+    """
+    The GreedyPolicy will always choose the optimal action.
+    """
+    def compute_action(self, obs, **kwargs):
+        return np.argmax(self.q_table[obs])
+
+    def probability(self, obs, action, **kwargs):
+        return 1 if action == np.argmax(self.q_table[obs]) else 0
+
+
+class EpsilonSoftPolicy(GreedyPolicy):
```

Comment:
The EpsilonPolicy becomes the GreedyPollicy when `epsilon` is 0. So maybe we should just combine these two into a single class.